### PR TITLE
Fix #1166: Support the portspan endpoint option in DIOP

### LINKED ISSUE #1166: DIOP endpoint port span support
Using TAO 2.5.5, we observed that using `portspan` with DIOP results in an error "Invalid DIOP option: <portspan>." I confirmed in source code that the `DIOP_Acceptor` does not support `portspan` in the latest master. We need to use the `portspan` endpoint option with DIOP, so we have modified the `DIOP_Acceptor` to support the `portspan` endpoint option (pull request coming after I file this issue).

### PR DESCRIPTION
**Pull request by nickwilliams-zaxiom**
- First, fixed the `ORB_init/Portspan` test, which was failing in master:
  - `Spawn()` returns instantly and does not wait for the process to fail, so the test was failing with the error "ERROR: Last server didn't fail! Err:0."
  - Updated the test to, after spawning, wait a second to make sure the process is still running (or, in the case of the last server, to check that the process has failed).
  - The test now passed.
- Next, updated `ORB_init/Portspan` to test DIOP as well as IIOP. The test then failed.
- Finally, updated `DIOP_Acceptor` to support `portspan` in the same manner as `IIOP_Acceptor` (namely, to loop through the port span until an open port is found). The test now passes again.

### CODE REVIEW

**Comment by jwillemsen** [comment]
Looks the test was not listed in https://github.com/DOCGroup/ACE_TAO/blob/master/TAO/bin/tao_orb_tests.lst, please add it there

**Comment by jwillemsen** [comment]
New option needs to be documented in the TAO documentation under TAO/docs

File: TAO/tao/Strategies/DIOP_Acceptor.cpp [hunk 1]
```diff
@@ -1038,6 +1071,20 @@ TAO_DIOP_Acceptor::parse_options (const char *str)
                                  ACE_TEXT ("endpoint priorities no longer supported.\n")),
                                 -1);
             }
+          else if (name == "portspan")
+            {
+              int range = static_cast <int> (ACE_OS::atoi (value.c_str ()));
+              // @@ What's the lower bound on the range?  zero, or one?
+              if (range < 1 || range > ACE_MAX_DEFAULT_PORT)
+                TAOLIB_ERROR_RETURN ((LM_ERROR,
+                                   ACE_TEXT ("TAO (%P|%t) Invalid IIOP endpoint ")
```
Comment (jwillemsen) [inline TAO/tao/Strategies/DIOP_Acceptor.cpp:1080]: DIOP

**Comment by jwillemsen** [comment]
test gets a dependency on strategies library which means the mpc file needs an update for a static build and the build order

**Comment by jwillemsen** [comment]
Would suggest to move the test one directory higher so that it is not hidden within the ORB init directory

File: TAO/tao/Strategies/DIOP_Acceptor.cpp (hunk 1 above)
Comment (nickwilliams-zaxiom) [reply TAO/tao/Strategies/DIOP_Acceptor.cpp:1080]: Fixed. My change will show when I push the next commit.

**Comment by nickwilliams-zaxiom** [comment]
> Looks the test was not listed in https://github.com/DOCGroup/ACE_TAO/blob/master/TAO/bin/tao_orb_tests.lst, please add it there. There is a NO_DIOP to disable DIOP related tests in this file when DIOP is not available, that seems not possible anymore to do with your test change, maybe make the protocol a commandline option and list the test twice in the lst file?

> Would suggest to move the test one directory higher so that it is not hidden within the ORB init directory

> New option needs to be documented in the TAO documentation under TAO/docs

> test gets a dependency on strategies library which means the mpc file needs an update for a static build and the build order

Changes applied as requested. They will show when I push the next commit.

**Comment by nickwilliams-zaxiom** [comment]
Okay, I have pushed a new commit that I believe addresses all of the comments.

File: TAO/tao/Strategies/DIOP_Acceptor.cpp [hunk 2]
```diff
@@ -1038,6 +1071,20 @@ TAO_DIOP_Acceptor::parse_options (const char *str)
                                  ACE_TEXT ("endpoint priorities no longer supported.\n")),
                                 -1);
             }
+          else if (name == "portspan")
+            {
+              int range = static_cast <int> (ACE_OS::atoi (value.c_str ()));
```
Comment (jwillemsen) [inline TAO/tao/Strategies/DIOP_Acceptor.cpp:1076]: could be const

File: TAO/tao/Strategies/DIOP_Acceptor.cpp [hunk 3]
```diff
@@ -393,27 +394,59 @@ int
 TAO_DIOP_Acceptor::open_i (const ACE_INET_Addr& addr,
                            ACE_Reactor *reactor)
 {
-  ACE_NEW_RETURN (this->connection_handler_,
-                  TAO_DIOP_Connection_Handler (this->orb_core_),
-                  -1);
+  unsigned short const requested_port = addr.get_port_number ();
+  ACE_UINT32 last_port = requested_port + this->port_span_ - 1;
+  if (last_port > ACE_MAX_DEFAULT_PORT)
```
Comment (jwillemsen) [inline TAO/tao/Strategies/DIOP_Acceptor.cpp:399]: Could use ACE_MIN here which could make last_port const

File: TAO/tao/Strategies/DIOP_Acceptor.cpp (hunk 2 above)
Comment (jwillemsen) [reply TAO/tao/Strategies/DIOP_Acceptor.cpp:1076]: as ACE_OS::atoi returns int a cast is not necessary here

**Comment by jwillemsen** [comment]
See #1220 for some related cleanup

File: TAO/docs/ORBEndpoint.html [hunk 4]
```diff
@@ -578,20 +578,54 @@ <h3>DIOP Endpoint Overview</h3>
...
+          <TH>Option</TH>
+          <TH>Availability</TH>
+          <TH>Description</TH>
+        </TR>
+        <TR>
+          <TD>
+            <CODE>portspan</CODE>
+          </TD>
+          <TD>
+            <CODE>TAO 2.5.11</CODE>
```
Comment (jwillemsen) [inline TAO/docs/ORBEndpoint.html:605]: Needs to be updated

File: TAO/tao/Strategies/DIOP_Acceptor.cpp (hunk 2 above)
Comment (jwillemsen) [inline TAO/tao/Strategies/DIOP_Acceptor.cpp:1076]:
``` suggestion
              int const range = ACE_OS::atoi (value.c_str ());

**Review by jwillemsen — CHANGES_REQUESTED** [review]

File: TAO/docs/ORBEndpoint.html (hunk 4 above)
Comment (nickwilliams-zaxiom) [reply TAO/docs/ORBEndpoint.html]: Updated.

File: TAO/tao/Strategies/DIOP_Acceptor.cpp (hunk 3 above)
Comment (nickwilliams-zaxiom) [reply TAO/tao/Strategies/DIOP_Acceptor.cpp]: Great suggestion. Done.

File: TAO/tao/Strategies/DIOP_Acceptor.cpp (hunk 2 above)
Comment (nickwilliams-zaxiom) [reply TAO/tao/Strategies/DIOP_Acceptor.cpp]: Good to know. Done per your suggestion below.

Comment (nickwilliams-zaxiom) [reply TAO/tao/Strategies/DIOP_Acceptor.cpp]: Implemented your suggestion.

File: TAO/tao/Strategies/DIOP_Acceptor.cpp (hunk 3 above)
Comment (jwillemsen) [reply TAO/tao/Strategies/DIOP_Acceptor.cpp]: Can you also update SCIOP_Acceptor.cpp and SSLIOP_Acceptor with the similar change?

Comment (nickwilliams-zaxiom) [reply TAO/tao/Strategies/DIOP_Acceptor.cpp]: Done.

**Review by jwillemsen — APPROVED** [review]

**Comment by jwillemsen** [comment]
Thanks!

**Pull request merged**